From bd89e93d6d7109971d3a8458f5734a23ae737886 Mon Sep 17 00:00:00 2001
From: Xianzhong <xianzhong.li@nxp.com>
Date: Thu, 6 Sep 2018 23:13:14 +0800
Subject: [PATCH 4594/5242] MGS-3940-3 gpu: imx: dpu-blit: fix prg bypass
 issue

commit  ae69c9edd9c2a126d80784ee1021c45ec6eb354e from
https://source.codeaurora.org/external/imx/linux-imx.git

render error happen when move mouse on chrome browser,
this can reproduce with alternant linear and tile blit.

the problem is related with prg reset for bypass switch,
a delay is required before reset prg, will check reason.

adding 10us delay before prg reset can fix bypass issue.
also reduce bitter delay to 10us to improve performance.

Signed-off-by: Xianzhong <xianzhong.li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dpu-blit/dpu-blit.c |    4 ++--
 drivers/gpu/imx/imx8_dprc.c         |    7 ++++++-
 drivers/gpu/imx/imx8_prg.c          |   10 +++++++++-
 3 files changed, 17 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/imx/dpu-blit/dpu-blit.c b/drivers/gpu/imx/dpu-blit/dpu-blit.c
index f33d163..4477890 100644
--- a/drivers/gpu/imx/dpu-blit/dpu-blit.c
+++ b/drivers/gpu/imx/dpu-blit/dpu-blit.c
@@ -42,7 +42,7 @@ static void dpu_cs_wait_fifo_space(struct dpu_bliteng *dpu_be)
 {
 	while ((dpu_be_read(dpu_be, CMDSEQ_STATUS) &
 		CMDSEQ_STATUS_FIFOSPACE_MASK) < CMDSEQ_FIFO_SPACE_THRESHOLD)
-		usleep_range(1000, 2000);
+		usleep_range(10, 20);
 }
 
 static void dpu_cs_wait_idle(struct dpu_bliteng *dpu_be)
@@ -246,7 +246,7 @@ void dpu_be_wait(struct dpu_bliteng *dpu_be)
 
 	while ((dpu_be_read(dpu_be, COMCTRL_INTERRUPTSTATUS0) &
 		STORE9_SEQCOMPLETE_IRQ_MASK) == 0)
-		usleep_range(1000, 2000);
+		usleep_range(10, 20);
 
 	dpu_be_write(dpu_be, STORE9_SEQCOMPLETE_IRQ_MASK,
 		COMCTRL_INTERRUPTCLEAR0);
diff --git a/drivers/gpu/imx/imx8_dprc.c b/drivers/gpu/imx/imx8_dprc.c
index c267824..46aa84e 100644
--- a/drivers/gpu/imx/imx8_dprc.c
+++ b/drivers/gpu/imx/imx8_dprc.c
@@ -256,7 +256,12 @@ static inline void dprc_write(struct dprc *dprc, u32 value, unsigned int offset)
 static void dprc_reset(struct dprc *dprc)
 {
 	dprc_write(dprc, SOFT_RESET, SYSTEM_CTRL0 + SET);
-	usleep_range(1000, 2000);
+
+	if (dprc->is_blit_chan)
+		usleep_range(10, 20);
+	else
+		usleep_range(1000, 2000);
+
 	dprc_write(dprc, SOFT_RESET, SYSTEM_CTRL0 + CLR);
 }
 
diff --git a/drivers/gpu/imx/imx8_prg.c b/drivers/gpu/imx/imx8_prg.c
index a7b2ae9..b113483 100644
--- a/drivers/gpu/imx/imx8_prg.c
+++ b/drivers/gpu/imx/imx8_prg.c
@@ -92,8 +92,16 @@ static inline void prg_write(struct prg *prg, u32 value, unsigned int offset)
 
 static void prg_reset(struct prg *prg)
 {
+	if (prg->is_blit)
+		usleep_range(10, 20);
+
 	prg_write(prg, SOFTRST, PRG_CTRL + SET);
-	usleep_range(1000, 2000);
+
+	if (prg->is_blit)
+		usleep_range(10, 20);
+	else
+		usleep_range(1000, 2000);
+
 	prg_write(prg, SOFTRST, PRG_CTRL + CLR);
 }
 
-- 
1.7.9.5

